TST use global_random_seed in sklearn/cluster/tests/test_spectral.py

irene000 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
#22827

#### What does this implement/fix? Explain your changes.
Use global_random_seed in tests of test_spectral.py

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

irene000 committed:
TST use global_random_seed in sklearn/cluster/tests/test_spectral.py

irene000 committed:
Merge branch 'main' into test_spectral_global_random_seed

betatim reviewed (approved):
Thanks for the PR. Looks good to me

cmarmo commented:
Hi @irene000 , thanks for your pull request!
Do you mind [running black](https://scikit-learn.org/dev/developers/contributing.html#pull-request-checklist)?
```
$ black sklearn/cluster/tests/test_spectral.py
```
This will fix the lint issue and trigger all the other checks. Thanks!

irene000 committed:
reformat code to black

irene000 committed:
Merge branch 'test_spectral_global_random_seed' of https://github.com/irene000/scikit-learn into test_spectral_global_random_seed

irene000 committed:
Merge branch 'main' into test_spectral_global_random_seed

irene000 committed:
reformat code to black

irene000 committed:
Merge branch 'test_spectral_global_random_seed' of https://github.com/irene000/scikit-learn into test_spectral_global_random_seed

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/irene000/24802

jeremiedbb committed:
[all random seeds]
test_spectral_clustering
test_spectral_clustering_sparse
test_precomputed_nearest_neighbors_filtering
test_affinities
test_cluster_qr
test_cluster_qr_permutation_invariance
test_discretize
test_spectral_clustering_with_arpack_amg_solvers
test_n_components

irene000 committed:
Merge branch 'main' into test_spectral_global_random_seed

jeremiedbb committed:
Merge remote-tracking branch 'upstream/main' into pr/irene000/24802

jeremiedbb committed:
adjust [all random seed]
test_spectral_clustering
test_spectral_clustering_sparse
test_precomputed_nearest_neighbors_filtering
test_affinities
test_cluster_qr
test_cluster_qr_permutation_invariance
test_discretize
test_spectral_clustering_with_arpack_amg_solvers
test_n_components

jeremiedbb reviewed (approved):
LGTM. Thanks @irene000

jeremiedbb committed:
Merge branch 'main' into test_spectral_global_random_seed

jeremiedbb commented:
For the record I slightly modified the dataset in one of the tests because it was failing for a single seed and after a deep dive into the algorithm it turns out that there's inherent randomness in scipy's `eigsh` which doesn't produce the same results each time with the same args. It happens in rare situations, probably in near degenerate settings, for instance here the eigen values were close to machine precsion.

I suspect that it's just the result of rouding errors that accumulate differently in different calls because of the multi-threaded underlying BLAS/LACPACK calls, in which case it's kind of expected and there's nothing to do about it.
